### ISSUE TITLE
MessageId#compare is not symmetric (MessageIdImpl, BatchMessageIdImpl)

### ISSUE THREAD

**Issue by racorn** [apache/pulsar#6433]
**Describe the bug**
Given to `MessageId`s, of which one happens to be a `MessageIdImpl` and the other is a `BatchMessageIdImpl`, the compare method is not symmetric.

**To Reproduce**
```
import static org.assertj.core.api.Assertions.assertThat;
@Test
public void shouldBeSymmetric() {

    MessageIdImpl simpleMessageId = new MessageIdImpl(1, 2, 3);
    BatchMessageIdImpl batchMessageId = new BatchMessageIdImpl(1, 2, 3, 0);
    assertThat(simpleMessageId.compareTo(batchMessageId)).isEqualTo(batchMessageId.compareTo(simpleMessageId));

    }
```

**Expected behavior**
MessageId#compare should be symmetric

**Additional context**
Pulsar 2.5.0

**Comment by jiazhai** [comment]
Thanks @racorn for the reporting. Would you like to provide a fix for this issue?

**Comment by racorn** [comment]
Yes, I could, but I am not sure how `MessageIdImpl` and `BatchMessageIdImpl` should compare?

It seems to me that the asymmetry was introduced by purpose with #1285, the code reads as
```
} else if (o instanceof MessageIdImpl) {
    int res = super.compareTo(o);
    if (res == 0 && batchIndex > NO_BATCH) {
        return 1;
    } else {
        return res;
    }
}
```
But as it is now, it breaks the `Comparable` contract, and that makes it confusing for user of the client API.

**Comment by zhanghaou** [comment]
I'm interesting in fix this issue.

**Comment by codelipenghui** [comment]
@zhanghaou Thanks, look forwards to your PR.

**Comment by tisonkun** [comment]
Closed as fixed.

Also, the original issue is incorrect because switching the operator of a corporation is never symmetric. The invariant here is:

```java
a.compareTo(b) + b.compareTo(a) == 0;
```

**Comment by racorn** [comment]
@tisonkun It is correct that in general `compareTo` is not symmetric, but `a.compareTo(b) == 0` implies that `b.compareTo(a)` should be 0 as well.

Javadoc for Comparable.compareTo states

**The implementor must ensure sgn(x.compareTo(y)) == -sgn(y.compareTo(x))for all x and y. (This implies that x.compareTo(y) must throw an exception iff y.compareTo(x) throws an exception.)** 

If I remember correctly, that requirement was not fulfilled for the test data in the original description.

**Comment by tisonkun** [comment]
@racorn Anyway, the compare logic for both classes is correct now :)

Some code quality improvements can be made, and you're welcome to give it a review or suggest a test case to cover this issue. See https://github.com/apache/pulsar/pull/18843.
